Allow to use private and protected parameterized constructor for mapping

ImoutoChan opened the pull request:
This PR adds an ability for linq2db to map data to classes with only private or protected constructors. Before this PR only public ones were supported. 

This case is useful when you want to map data to encapsulated rich models, which often have closed constructors.

We can also consider smartly selecting which constructor to use (most suitable based on field names). Right now it will only work for classes with single constructor.

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/linq2db/linq2db?pullRequest=3147) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/linq2db/linq2db?pullRequest=3147) before we can accept your contribution.<br/><br/>:x: ImoutoChan [sign now](https://cla.dotnetfoundation.org/linq2db/linq2db?pullRequest=3147)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/linq2db/linq2db?pullRequest=3147) it.</sub>

sdanyliv commented:
/azp run test-sqlite

ImoutoChan commented:
What do you think? Can it be merged?

Or did I break something? I don't see which test is broken.

MaceWindu commented:
/azp run test-all

MaceWindu commented:
Thanks!

It's failed because we don't support db-specific tests on external pr, so I need to run all tests.

Regarding implementation, I think it should be changed a bit - first probe public constructors and if there is not match then try non-public. Otherwise it could be a regression for existing types with public and non-public constructors

MaceWindu reviewed (changes requested):
see my comment above

ImoutoChan commented:
Thank you for your feedback!

I've changed implementation as suggested to try pubic constructor first. I've added one more test to check this specific case. The failed test should also be fixed now.

sdanyliv commented:
/azp run test-all
